### PR TITLE
Query reactions in Ruby; add/remove reactions in Java and Ruby

### PR DESCRIPTION
**Pull request by johnkerl**
## Test add/remove reactions in Ruby, using Ruby to inspect

Two apps, single threat descriptor, initially no reactions by either of them.

```
$ te-tag-query-ruby -a TX_ACCESS_TOKEN_MHST ids-to-details 2589692907727086 | jq .reactions

null
```

First app reacts

```
$ te-tag-query-ruby -a TX_ACCESS_TOKEN_MHST update --reactions-to-add HELPFUL -i 2589692907727086
{"success":true}

$ te-tag-query-ruby -a TX_ACCESS_TOKEN_MHST ids-to-details 2589692907727086 | jq .reactions
[
  {
    "key": "HELPFUL",
    "value": "1064060413755420"
  }
]
```

Second app reacts

```
$ te-tag-query-ruby -a TX_ACCESS_TOKEN_MHSRFT update --reactions-to-add HELPFUL,INGESTED -i 2589692907727086
{"success":true}

$ te-tag-query-ruby -a TX_ACCESS_TOKEN_MHST ids-to-details 2589692907727086 | jq .reactions
[
  {
    "key": "HELPFUL",
    "value": "494491891138576,1064060413755420"
  },
  {
    "key": "INGESTED",
    "value": "494491891138576"
  }
]
```

First app unreacts once

```
$ te-tag-query-ruby -a TX_ACCESS_TOKEN_MHSRFT update --reactions-to-remove HELPFUL -i 2589692907727086
{"success":true}

$ te-tag-query-ruby -a TX_ACCESS_TOKEN_MHST ids-to-details 2589692907727086 | jq .reactions
[
  {
    "key": "HELPFUL",
    "value": "1064060413755420"
  },
  {
    "key": "INGESTED",
    "value": "494491891138576"
  }
]
```

Second app unreacts

```
$ te-tag-query-ruby -a TX_ACCESS_TOKEN_MHST update --reactions-to-remove HELPFUL -i 2589692907727086
{"success":true}

$ te-tag-query-ruby -a TX_ACCESS_TOKEN_MHST ids-to-details 2589692907727086 | jq .reactions
[
  {
    "key": "INGESTED",
    "value": "494491891138576"
  }
]
```

First app unreacts a second time

```
$ te-tag-query-ruby -a TX_ACCESS_TOKEN_MHSRFT update --reactions-to-remove INGESTED -i 2589692907727086
{"success":true}

$ te-tag-query-ruby -a TX_ACCESS_TOKEN_MHST ids-to-details 2589692907727086 | jq .reactions
null
```

## Test add/remove reactions in Java, using Ruby to inspect

Two apps, single threat descriptor, initially no reactions by either of them.

```
$ ../te-tag-query-ruby/te-tag-query-ruby -a TX_ACCESS_TOKEN_MHST ids-to-details 2589692907727086 | jq .reactions

null
```

```
$ alias jtq='java com.facebook.threatexchange.TETagQuery'
```

First app reacts

```
$ jtq -a TX_ACCESS_TOKEN_MHST update --reactions-to-add HELPFUL -i 2589692907727086
{"success":true}

$ ../te-tag-query-ruby/te-tag-query-ruby -a TX_ACCESS_TOKEN_MHST ids-to-details 2589692907727086 | jq .reactions
[
  {
    "key": "HELPFUL",
    "value": "1064060413755420"
  }
]
```

Second app reacts

```
$ jtq -a TX_ACCESS_TOKEN_MHSRFT update --reactions-to-add HELPFUL,INGESTED -i 2589692907727086
{"success":true}

$ ../te-tag-query-ruby/te-tag-query-ruby -a TX_ACCESS_TOKEN_MHST ids-to-details 2589692907727086 | jq .reactions
[
  {
    "key": "HELPFUL",
    "value": "494491891138576,1064060413755420"
  },
  {
    "key": "INGESTED",
    "value": "494491891138576"
  }
]
```

First app unreacts once

```
$ jtq -a TX_ACCESS_TOKEN_MHSRFT update --reactions-to-remove HELPFUL -i 2589692907727086
{"success":true}

$ ../te-tag-query-ruby/te-tag-query-ruby -a TX_ACCESS_TOKEN_MHST ids-to-details 2589692907727086 | jq .reactions
[
  {
    "key": "HELPFUL",
    "value": "1064060413755420"
  },
  {
    "key": "INGESTED",
    "value": "494491891138576"
  }
]
```

Second app unreacts

```
$ jtq -a TX_ACCESS_TOKEN_MHST update --reactions-to-remove HELPFUL -i 2589692907727086
{"success":true}

$ ../te-tag-query-ruby/te-tag-query-ruby -a TX_ACCESS_TOKEN_MHST ids-to-details 2589692907727086 | jq .reactions
[
  {
    "key": "INGESTED",
    "value": "494491891138576"
  }
]
```

First app unreacts a second time

```
$ jtq -a TX_ACCESS_TOKEN_MHSRFT update --reactions-to-remove INGESTED -i 2589692907727086
{"success":true}

$ ../te-tag-query-ruby/te-tag-query-ruby -a TX_ACCESS_TOKEN_MHST ids-to-details 2589692907727086 | jq .reactions
null
```

### CODE REVIEW

File: hashing/te-tag-query-java/com/facebook/threatexchange/TETagQuery.java [hunk 1]
```diff
@@ -945,15 +945,19 @@ public void usage(int exitCode) {
...
       o.printf("-r|--review-status {...}\n");
       o.printf("--first-active {...}\n");
       o.printf("--last-active {...}\n");
       o.printf("--expired-on {...}\n");
       o.printf("\n");
-      o.printf("Please see the following for allowed values in all enumerated types:\n");
+      o.printf("Please see the following for allowed values in all enumerated types except reactions:\n");
       o.printf("https://developers.facebook.com/docs/threat-exchange/reference/submitting\n");
+      o.printf("Please see the following for allowed values in all enumerated types in reactions:\n");
+      o.printf("https://developers.facebook.com/docs/threat-exchange/reference/reacting\n");
```
Comment (Dcallies) [inline hashing/te-tag-query-java/com/facebook/threatexchange/TETagQuery.java:960]: Hmm, this definitely makes me miss python libraries for CLIs like argparse and click.

File: hashing/te-tag-query-ruby/TETagQuery.rb [hunk 2]
```diff
@@ -585,22 +585,32 @@ def usage(exitCode)
...
 --expired-on {...}
 
-Please see the following for allowed values in all enumerated types:
+Please see the following for allowed values in all enumerated types except reactions:
 https://developers.facebook.com/docs/threat-exchange/reference/submitting
 
+Please see the following for enumerated types in reactions:
+See also https://developers.facebook.com/docs/threat-exchange/reference/reacting
+
+xxx reactions too
```
Comment (Dcallies) [inline hashing/te-tag-query-ruby/TETagQuery.rb:612]: Is this a missed TODO?

Comment (johnkerl) [reply hashing/te-tag-query-ruby/TETagQuery.rb:612]: good catch -- thanks!

File: hashing/te-tag-query-java/com/facebook/threatexchange/TETagQuery.java (hunk 1 above)
Comment (johnkerl) [reply hashing/te-tag-query-java/com/facebook/threatexchange/TETagQuery.java:960]: There is definitely `getopt` ... there's a note elsewhere in the file about why i'm not using it here (fails to separate main-flags from verb-flags). For any CLI programs without verb-handlers, `getopt` is a great go-to, and (as a bonus) works largely the same across several languages.

File: hashing/te-tag-query-ruby/TETagQuery.rb (hunk 2 above)
Comment (johnkerl) [reply hashing/te-tag-query-ruby/TETagQuery.rb:612]: I'll do this very next commit (promise!) :)

I have another diff in my local repo one ahead of this & my git-fu is weaker than my hg-fu.

Comment (johnkerl) [reply hashing/te-tag-query-ruby/TETagQuery.rb:612]: https://github.com/facebook/ThreatExchange/commit/654836ca119a65324d7726a245c155592b458f8e

**Pull request merged**